ampd/src/handlers/solana_verify_worker_set.rs

```diff
@@ -0,0 +1,415 @@
+#[try_from("wasm-worker_set_poll_started")]
+struct PollStartedEvent {
+    #[serde(rename = "_contract_address")]
+    contract_address: TMAddress,
+    worker_set: WorkerSetConfirmation,
+    poll_id: PollId,
+    source_chain: connection_router::state::ChainName,
+    source_gateway_address: String,
+    expires_at: u64,
+    confirmation_height: u64,
+    participants: Vec<TMAddress>,
+}
+
+pub struct Handler<B>
+where
+    B: BroadcasterClient,
+{
+    worker: TMAddress,
+    voting_verifier: TMAddress,
+    chain: ChainName,
+    rpc_client: RpcClient,
+    broadcast_client: B,
+    latest_block_height: Receiver<u64>,
+}
+
+impl<B> Handler<B>
+where
+    B: BroadcasterClient,
+{
+    pub fn new(
+        worker: TMAddress,
+        voting_verifier: TMAddress,
+        chain: ChainName,
+        rpc_client: RpcClient,
+        broadcast_client: B,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        Self {
+            worker,
+            voting_verifier,
+            chain,
+            rpc_client,
+            broadcast_client,
+            latest_block_height,
+        }
+    }
+
+    async fn broadcast_vote(&self, poll_id: PollId, vote: Vote) -> Result<()> {
+        let msg = serde_json::to_vec(&ExecuteMsg::Vote {
+            poll_id,
+            votes: vec![vote],
+        })
+        .expect("vote msg should serialize");
+        let tx = MsgExecuteContract {
+            sender: self.worker.as_ref().clone(),
+            contract: self.voting_verifier.as_ref().clone(),
+            msg,
+            funds: vec![],
+        };
+
+        self.broadcast_client
+            .broadcast(tx)
+            .await
+            .change_context(Error::Broadcaster)
+    }
+}
+
+#[async_trait]
+impl<B> EventHandler for Handler<B>
+where
+    B: BroadcasterClient + Send + Sync,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &events::Event) -> Result<()> {
+        let PollStartedEvent {
+            contract_address,
+            poll_id,
+            source_chain,
+            source_gateway_address,
+            expires_at,
+            confirmation_height: _,
+            participants,
+            worker_set,
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(())
+            }
+            event => event.change_context(Error::DeserializeEvent)?,
+        };
+
+        if self.voting_verifier != contract_address {
+            return Ok(());
+        }
+
+        if self.chain != source_chain {
+            return Ok(());
+        }
+
+        if !participants.contains(&self.worker) {
+            return Ok(());
+        }
+
+        let latest_block_height = *self.latest_block_height.borrow();
+        if latest_block_height >= expires_at {
+            info!(poll_id = poll_id.to_string(), "skipping expired poll");
+            return Ok(());
+        }
```

Comment:
From the docs:

> If the receiver is only interested in the current value, and does not intend to wait for changes, then [Receiver::borrow()](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.borrow) can be used. It may be more convenient to use [borrow](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.borrow) since it’s an &self method—[borrow_and_update](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.borrow_and_update) requires &mut self.

Isnt it our case here ? - this event handler just needs to get the value at that time and not waiting for any other message.